aten/src/ATen/native/Sorting.cpp

```diff
@@ -80,20 +83,125 @@ void quick_select_template(
       if (j < i)
         break;
       swap_fn(i, j);
-    } while (1);
+    } while (true);
     swap_fn(L, j);
 
     // Re-set active partition
     if (j <= k)
       L = i;
     if (j >= k)
       R = j - 1;
-  } while (1);
+  } while (true);
 }
 
-} // namespace
+void quantile_impl(
+    Tensor& out,
+    const Tensor& self,
+    const Tensor& q,
+    optional<int64_t> _dim,
+    bool keepdim) {
+  int64_t dim = at::maybe_wrap_dim(_dim.value_or(0), self.dim(), true);
+
+  TORCH_CHECK(self.numel() > 0, "quantile() input tensor must be non-empty");
+  TORCH_CHECK(q.dim() <= 1, "quantile() q must be a scalar or 1D tensor");
+
+  TORCH_CHECK(
+      self.scalar_type() == kFloat || self.scalar_type() == kDouble,
+      "quantile() input tensor must be either float or double dtype");
+  TORCH_CHECK(
+      self.scalar_type() == q.scalar_type(),
+      "quantile() q must be same dtype as the input tensor");
+  TORCH_CHECK(
+      self.device() == q.device(),
+      "quantile() q must be on the same device as the input tensor");
+
+  // Compute output shape: q_size + reduced_size
+  std::vector<int64_t> out_shape;
+  if (_dim && self.dim() > 0) {
+    out_shape = self.sizes().vec();
+    if (keepdim) {
+      out_shape[dim] = 1;
+    } else {
+      out_shape.erase(out_shape.begin() + dim);
+    }
+  } else if (keepdim) {
+    out_shape = std::vector<int64_t>(self.dim(), 1);
+  }
+  if (q.dim() > 0) {
+    out_shape.insert(out_shape.begin(), q.numel());
+  }
+
+  // Check or resize output
+  if (out.numel() > 0) {
+    TORCH_CHECK(
+        out.sizes().vec() == out_shape,
+        "quantile() expected output shape to be ",
+        out_shape,
+        " but got ",
+        out.sizes().vec());
+    TORCH_CHECK(
+        self.scalar_type() == out.scalar_type(),
+        "quantile() out tensor must be same dtype as the input tensor");
+    TORCH_CHECK(
+        self.device() == out.device(),
+        "quantile() out tensor must be on the same device as the input tensor");
+  } else {
+    resize_output(out, out_shape);
+  }
+
+  // Do this check only on cpu to prevent implicit sync on CUDA
```

Comment:
```
// Checks that all q values are between 0 and 1, inclusive
// NOTE: this check is only performed when running on the CPU to avoid synchronizing the CPU with an accelerator
```